design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+# KLIP 12 - Implement High-Availability for Pull queries
+
+**Author**: Vicky Papavasileiou, Vinoth Chandar | 
+**Release Target**: 5.5 | 
+**Status**: _In Discussion_
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** Enables high-availability for Ksql pull queries, in case of server failures. Current 
+design for handling failure of an active ksql server, incurs an unavailability period of several 
+seconds (10+) to few minutes due to the Streams rebalancing procedure. This work is based on new 
```

Comment:
What causes it to take so long?  Does rebalancing unsure that the replica is fully caught up to the former active, ensuring consistency, before it can become the new active?

Reply:
Here is some background information on how rebalancing works and what is the current/future work to optimize it https://cwiki.apache.org/confluence/display/KAFKA/KIP-429%3A+Kafka+Consumer+Incremental+Rebalance+Protocol

Reply:
> Does rebalancing unsure that the replica is fully caught up to the former active, ensuring consistency, before it can become the new active?

Correct. KIP-429 does not really change this behavior. However, KIP-441 (https://cwiki.apache.org/confluence/display/KAFKA/KIP-441%3A+Smooth+Scaling+Out+for+Kafka+Streams) would address this issue.